### PR TITLE
added API security on Nuts node using asymetric keys

### PR DESCRIPTION
**Pull request by woutslakhorst**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,6 +38,8 @@ $ docker run -p 1303:1303 nutsfoundation/nuts-registry-admin-demo
 When running in Docker without a config file mounted at `/app/server.config.yaml` it will use the default configuration.
 In this case the default username will be `demo@nuts.nl`. The password is generated and printed in the log on startup.
 
+The `nutsnodeapikeyfile` config parameter should point to a PEM encoded private key file. The corresponding public key should be configured on the Nuts node in SSH authorized keys format.
```
Comment (reinkrul) [inline README.md:41]: These "nutsnode" prefixes to the config keys are a bit... ugly. But I guess consistency is better for now.

Also document `nutsnodeapiuser`

File: Dockerfile [hunk 2]
```diff
@@ -12,7 +12,7 @@ RUN npm run build
 #
 # Build backend
 #
-FROM golang:1.17-alpine as backend-builder
+FROM golang:1.19-alpine as backend-builder
```
Comment (reinkrul) [inline Dockerfile:15]: why not 1.20?

File: config.go [hunk 3]
```diff
@@ -118,6 +128,18 @@ func loadConfig() Config {
 		log.Fatalf("error while unmarshalling config: %v", err)
 	}
 
+	// Load the API key
```
Comment (reinkrul) [inline config.go:131]: I think `nutsnodeapiuser` is required now, so can lead to an error when nog configured

File: config.go [hunk 4]
```diff
@@ -156,3 +178,23 @@ func envProvider() *env.Env {
 			strings.TrimPrefix(s, defaultPrefix)), "_", defaultDelimiter, -1)
 	})
 }
+
+// pemToPrivateKey converts a PEM encoded private key to a Signer interface. It supports EC, RSA and PKIX PEM encoded strings
+func pemToPrivateKey(bytes []byte) (signer crypto.Signer, err error) {
+	key, _ := ssh.ParseRawPrivateKey(bytes)
```
Comment (reinkrul) [inline config.go:184]: why from the `ssh` package? Isn't it a normal EC/RSA PKCS-something encoded key?

File: go.mod [hunk 5]
```diff
@@ -1,161 +1,163 @@
...
-	github.com/mitchellh/mapstructure v1.4.3 // indirect
+	github.com/mitchellh/mapstructure v1.5.0 // indirect
 	github.com/mitchellh/reflectwalk v1.0.2 // indirect
 	github.com/mr-tron/base58 v1.1.3 // indirect
 	github.com/multiformats/go-multihash v0.0.11 // indirect
-	github.com/nats-io/jwt/v2 v2.2.1-0.20220330180145-442af02fd36a // indirect
-	github.com/nats-io/nats-server/v2 v2.8.1 // indirect
-	github.com/nats-io/nats.go v1.14.0 // indirect
+	github.com/nats-io/jwt/v2 v2.3.0 // indirect
+	github.com/nats-io/nats-server/v2 v2.9.14 // indirect
```
Comment (reinkrul) [inline go.mod:104]: so many irrelevant dependencies... 😢

File: main.go [hunk 6]
```diff
@@ -92,14 +100,30 @@ func main() {
 	}
 	auth := api.NewAuth(config.sessionKey, []api.UserAccount{account})
 
+	// API security
+	tokenGenerator := func() (string, error) {
+		return "", nil
+	}
+	if config.apiKey != nil {
+		// debug
```
Comment (reinkrul) [inline main.go:108]: ?

File: main.go [hunk 7]
```diff
@@ -181,3 +205,75 @@ func httpErrorHandler(err error, c echo.Context) {
...
+			return "", err
+		}
+
+		issuedAt := time.Now()
+		notBefore := issuedAt
+		expires := notBefore.Add(time.Second * time.Duration(5))
+		token, err := jwt.NewBuilder().
+			Issuer(config.NutsNodeAPIUser).
+			Subject(config.Credentials.Username).
+			Audience([]string{config.NutsNodeAddress}).
```
Comment (reinkrul) [inline main.go:223]: shouldn't this be the configured name? Or at least just the hostname from the address

File: main.go [hunk 8]
```diff
@@ -181,3 +205,75 @@ func httpErrorHandler(err error, c echo.Context) {
...
+
+	switch k := signer.(type) {
+	case *rsa.PrivateKey:
+		key.Set(jwk.AlgorithmKey, jwa.PS512)
+	case *ecdsa.PrivateKey:
+		var alg jwa.SignatureAlgorithm
+		alg, err = ecAlg(k)
+		key.Set(jwk.AlgorithmKey, alg)
+	default:
+		err = errors.New("unsupported signing private key")
```
Comment (reinkrul) [inline main.go:253]:
```suggestion
		err = fmt.Errorf("unsupported signing private key: %T", k)
```

File: server.config.yaml [hunk 9]
```diff
@@ -2,6 +2,8 @@ credentials:
   username: "demo@nuts.nl"
   password: "demo"
 port: 1303
-nutsNodeAddr: "http://localhost:1323"
+nutsnodeaddr: "http://localhost:1323"
+nutsnodeapikeyfile: "./skey.pem"
+nutsnodeapiuser: "demo@woutslakhorst.nl"
```
Comment (reinkrul) [inline server.config.yaml:7]: do you. want this checked in?

File: package-lock.json [hunk 10]
```diff
@@ -205,6 +205,64 @@
         "purgecss": "^3.1.3"
       }
     },
+    "node_modules/@jridgewell/gen-mapping": {
```
Comment (reinkrul) [inline package-lock.json:208]: why this file changed?

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 2 above)
Comment (woutslakhorst) [reply Dockerfile:15]: Still waiting for mac 1.20 version, but for builder this wouldn't matter....

File: config.go (hunk 4 above)
Comment (woutslakhorst) [reply config.go:184]: ssh covers all "normal" encoded keys and the SSH format. So it can handle more

File: main.go (hunk 7 above)
Comment (woutslakhorst) [reply main.go:223]: `aud` is specified to be a URI. The configured address is a URI. The node should have the correct value configured. Clients should use the address.

File: package-lock.json (hunk 10 above)
Comment (woutslakhorst) [reply package-lock.json:208]: When debugging, I ran a general npm build. So now it's updated to latest versions.

File: config.go [hunk 11]
```diff
@@ -156,3 +181,23 @@ func envProvider() *env.Env {
...
+		return
+	}
+
+	switch k := key.(type) {
+	case *rsa.PrivateKey:
+		signer = k
+	case *ecdsa.PrivateKey:
+		signer = k
+	default:
+		err = fmt.Errorf("unknown private key type: %s", k)
```
Comment (reinkrul) [inline config.go:199]:
```suggestion
		err = fmt.Errorf("unsupported private key type: %T", k)
```

**Review by reinkrul — CHANGES_REQUESTED** [review]

**Review by reinkrul — APPROVED** [review]

**Pull request merged**